KAFKA-17986: Fix ConsumerRebootstrapTest and ProducerRebootstrapTest

peterxcli opened the pull request:
`ProducerRebootstrapTest` has many issues in using kraft.

1. it passes `unclean.leader.election.enable` as server config but it is not configured on controller node. We should set that config to topic while creating topic.

2. ~~it assumes `topic` is distributed on [0, 1]. If the order is not correct, the test can't pass. We should create the topic manually to ensure the distribution.~~ The re-election is significantly small to the producer timeout, so I think we can ignore this here.

3. ~~in kraft mode, the time to trigger unclean election is 5 mins. it is too long to wait. We should configure controller to reduce the number.~~ `unclean.leader.election.interval.ms` is merely a periodic task that scans for topic partitions without a leader and, based on the value of `unclean.leader.election.enable`(topic's || cluster's), determines whether to perform unclean leader elections for those partitions.

peterxcli committed:
Fix rebootstrap test configuration (KAFKA-17986)

- Add unclean.leader.election.enable=true to topic configuration when creating
  test topic to ensure controller node has the correct setting
- Set unclean.leader.election.interval.ms to 1ms (down from default 5 minutes)
  to speed up test execution in KRaft mode

chia7712 commented:
@peterxcli I update the topic to match the changes. it is ok to file two PRs :)

peterxcli committed:
wip

TaiJuWu commented on core/src/test/java/kafka/server/ProducerRebootstrapIntegrationTest.java:
```diff
@@ -0,0 +1,267 @@
+            Properties topicProps = new Properties();
+            topicProps.put(TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, "true");
+
+            admin.createTopics(List.of(new NewTopic(topicName, 2, (short) 2).configs(Utils.propsToStringMap(topicProps)))).all().get();
```
Should we need `cluster.waitForTopic(topicName, 2);`  here?

peterxcli replied:
I think that's only necessary when the topic metadata has changed?

TaiJuWu replied:
Yes, but create topic will change metadata, right?

peterxcli replied:
Oh thanks, you're right. 
But seems the old test infra doesn't wait for the metadata propagation explicitly.
So I think it's ok the skip that.
https://github.com/apache/kafka/blob/772aa241b2f1c3e8c8b7ad73ddc910c8272f2d1d/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L380-L406

chia7712 commented on core/src/test/java/kafka/server/ProducerRebootstrapIntegrationTest.java:
```diff
@@ -0,0 +1,267 @@
+        brokers = 2,
+        serverProperties = {
+            @ClusterConfigProperty(key = TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, value = "true"),
+            @ClusterConfigProperty(key = ReplicationConfigs.UNCLEAN_LEADER_ELECTION_INTERVAL_MS_CONFIG, value = "1"),
```
this value is too small ... `3000` is good enough

peterxcli committed:
Add support for fixed broker ports in ClusterTest

Add ability to configure fixed broker ports in test framework instead of using ephemeral ports. This helps with:

- Debugging by making broker ports predictable and consistent
- Avoiding port conflicts in certain test scenarios
- Better control over broker port assignment

Changes include:
- Add fixedBrokerPorts flag to TestKitNodes and ClusterConfig
- Add port selection logic in KafkaClusterTestKit
- Add choosePort() utility method to TestUtils
- Update ClusterTest annotation to support fixed ports configuration

The change is backward compatible as fixedBrokerPorts defaults to false, maintaining existing behavior.

peterxcli committed:
Revert "Add support for fixed broker ports in ClusterTest"

This reverts commit 33e7c12f655efa97ca8ac0cd44c9672c0a6d1a0a.

peterxcli committed:
Revert "wip"

This reverts commit 5580f3edf63efb4eff88a7d4342763d906e50a77.

peterxcli committed:
fix(tests): re-enable ConsumerRebootstrapTest and improve comments

- Remove @Disabled annotation from ConsumerRebootstrapTest
- Add clarifying comment about leader shutdown in ProducerRebootstrapTest
- Improve documentation in RebootstrapTest regarding unclean leader election
- Clean up comment formatting

peterxcli commented:
This PR only fix these two tests, will migrate them to `ClusteTest` after I add the ability to use fixed broker port with it.

peterxcli commented:
Hi @chia7712, this PR is ready for review, PTAL. Thanks!

peterxcli commented:
> @peterxcli I update the topic to match the changes. it is ok to file two PRs :)

As we discussed https://github.com/apache/kafka/pull/18175#issuecomment-2548241448, change this PR title back to "Fix ConsumerRebootstrapTest and ProducerRebootstrapTest".

chia7712 commented on core/src/test/scala/integration/kafka/api/RebootstrapTest.scala:
```diff
@@ -20,18 +20,36 @@ import kafka.server.{KafkaBroker, KafkaConfig}
     overridingProps.put(GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, brokerCount.toString)
+    // Since DELIVERY_TIMEOUT_MS_CONFIG defaults to 120000ms,
+    // we need to set a smaller value for unclean.leader.election.interval.ms
+    overridingProps.put(ReplicationConfigs.UNCLEAN_LEADER_ELECTION_INTERVAL_MS_CONFIG, TimeUnit.MILLISECONDS.toMillis(1))
```
we should set this config for controller, right?

peterxcli committed:
Remove the wrongly added unclean election config in broker creating process

peterxcli commented:
I think my original thought was wrong

`unclean.leader.election.enable` has two scopes: topic-level and cluster-wide. If the cluster is configured with `unclean.leader.election.enable=true` at startup, the interval setting (`unclean.leader.election.interval.ms`) becomes irrelevant. This is because whenever the leader of a topic partition’s ISR is fenced, an unclean election will occur immediately. IMO, `unclean.leader.election.interval.ms` only become necessary when the cluster disable unclean election and enable it dynamically.

> `unclean.leader.election.interval.ms` is merely a periodic task that scans for topic partitions without a leader and, based on the value of `unclean.leader.election.enable`(topic's || cluster's), determines whether to perform unclean leader elections for those partitions.

---

Regarding enabling unclean elections at the cluster or topic level: Currently, in `RebootstrapTest`, only the relevant topics have unclean election enabled. I believe it is unnecessary to enable cluster-wide unclean election because those occurrence of  unclean election for some topics are expected.

peterxcli commented on core/src/test/scala/integration/kafka/api/ProducerRebootstrapTest.scala:
```diff
@@ -18,15 +18,14 @@ package kafka.api
   @ParameterizedTest(name = "{displayName}.quorum=kraft.useRebootstrapTriggerMs={0}")
   @ValueSource(booleans = Array(false, true))
   def testRebootstrap(useRebootstrapTriggerMs: Boolean): Unit = {
+    // It's ok to shut the leader down, cause the reelection is small enough to the producer timeout.
```
tested
```scala
val partitionAssigment = Map(0 -> Seq(1, 0), 1 -> Seq(1, 0))
createTopicWithAssignmentAndConfig(topic, partitionReplicaAssignment = partitionAssigment, topicConfig = topicProps)
```

and 

```scala
val partitionAssigment = Map(0 -> Seq(0 1), 1 -> Seq(0, 1))
createTopicWithAssignmentAndConfig(topic, partitionReplicaAssignment = partitionAssigment, topicConfig = topicProps)
```
as the topic's assignment, both of them worked well.

peterxcli replied:
helper function I added to test above stuff.
```scala
def createTopicWithAssignmentAndConfig(
 topic: String,
 partitionReplicaAssignment: collection.Map[Int, Seq[Int]],
 topicConfig: Properties = new Properties,
 listenerName: ListenerName = listenerName,
 adminClientConfig: Properties = new Properties
): scala.collection.immutable.Map[Int, Int] = {
  if (isKRaftTest()) {
    Using.resource(createAdminClient(brokers, listenerName, adminClientConfig)) { admin =>
      TestUtils.createTopicWithAdmin(
        admin = admin,
        topic = topic,
        brokers = brokers,
        controllers = controllerServers,
        replicaAssignment = partitionReplicaAssignment,
        topicConfig = topicConfig
      )
    }
  } else {
    TestUtils.createTopic(
      zkClient = zkClient,
      topic = topic,
      partitionReplicaAssignment = partitionReplicaAssignment,
      servers = servers,
      topicConfig = topicConfig
    )
  }
}
```

peterxcli commented:
Failed tests are tracked by:
- https://issues.apache.org/jira/browse/KAFKA-15474
